codegen/src/main/java/software/amazon/awssdk/codegen/jmespath/parser/JmesPathParser.java

```diff
@@ -0,0 +1,1090 @@
+        char unicodeChar;
+        try {
+            unicodeChar = (char) Integer.parseInt(unicodePattern, 16);
+        } catch (NumberFormatException e) {
+            logError("escaped-unicode-sequence", "Invalid unicode hex sequence", startPosition);
+            return ParseResult.error();
+        }
+
+        return ParseResult.success(String.valueOf(unicodeChar));
+    }
+
+    /**
+     * "*"
+     */
+    private ParseResult<WildcardExpression> parseWildcardExpression(int startPosition, int endPosition) {
+        return parseExpectedToken("star-expression", startPosition, endPosition, '*').mapResult(v -> new WildcardExpression());
+    }
+
+    private int charsInRange(int startPosition, int endPosition) {
+        return endPosition - startPosition;
+    }
+
+    private List<Integer> findCharacters(int startPosition, int endPosition, String symbol) {
+        List<Integer> results = new ArrayList<>();
+
+        int start = startPosition;
+        while (true) {
+            int match = input.indexOf(symbol, start);
+            if (match < 0 || match >= endPosition) {
+                break;
+            }
+            results.add(match);
+            start = match + 1;
+        }
+
+        return results;
+    }
+
+    private ParseResult<Character> parseExpectedToken(String parser, int startPosition, int endPosition, char expectedToken) {
+        if (input.charAt(startPosition) != expectedToken) {
+            logError(parser, "Expected '" + expectedToken + "'", startPosition);
+            return ParseResult.error();
+        }
+
+        if (charsInRange(startPosition, endPosition) != 1) {
+            logError(parser, "Unexpected character", startPosition + 1);
+            return ParseResult.error();
+        }
+
+        return ParseResult.success(expectedToken);
+    }
+
+    private int trimLeftWhitespace(int startPosition, int endPosition) {
+        while (input.charAt(startPosition) == ' ' && startPosition < endPosition - 1) {
+            ++startPosition;
+        }
+
+        return startPosition;
+    }
+
+    private int trimRightWhitespace(int startPosition, int endPosition) {
+        while (input.charAt(endPosition - 1) == ' ' && startPosition < endPosition - 1) {
+            --endPosition;
+        }
+
+        return endPosition;
+    }
+
+    private boolean startsWith(int startPosition, char character) {
+        return input.charAt(startPosition) == character;
+    }
+
+    private boolean endsWith(int endPosition, char character) {
+        return input.charAt(endPosition - 1) == character;
+    }
+
+    private boolean startsAndEndsWith(int startPosition, int endPosition, char startChar, char endChar) {
+        return startsWith(startPosition, startChar) && endsWith(endPosition, endChar);
+    }
+
+    private void logError(String parser, String message, int position) {
+        log.debug(() -> parser + " at " + position + ": " + message);
```

Comment:
It's an "error", but only in the context of an individual parser. The grammar we're implementing is very much a "try to parse the input, and if it can't be parsed, backtrack and try something else." 

This gets called every time we backtrack for the purposes of debugging "why did it backtrack?", but most of the time we don't care about these logs.